agent/workers/trigger.go

```diff
@@ -0,0 +1,219 @@
+package workers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kubeshop/tracetest/agent/client"
+	"github.com/kubeshop/tracetest/agent/proto"
+	agentTrigger "github.com/kubeshop/tracetest/agent/workers/trigger"
+	"github.com/kubeshop/tracetest/server/pkg/id"
+	"github.com/kubeshop/tracetest/server/test/trigger"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type TriggerWorker struct {
+	client   *client.Client
+	registry *agentTrigger.Registry
+}
+
+func NewTriggerWorker(client *client.Client) *TriggerWorker {
+	// TODO: use a real tracer
+	tracer := trace.NewNoopTracerProvider().Tracer("noop")
+
+	registry := agentTrigger.NewRegistry(tracer)
+	registry.Add(agentTrigger.HTTP())
+	registry.Add(agentTrigger.GRPC())
+	registry.Add(agentTrigger.TRACEID())
+
+	return &TriggerWorker{client, registry}
+}
+
+func (w *TriggerWorker) Trigger(ctx context.Context, triggerRequest *proto.TriggerRequest) error {
+	triggerConfig := convertProtoToTrigger(triggerRequest.Trigger)
+	triggerer, err := w.registry.Get(triggerConfig.Type)
+	if err != nil {
+		return err
+	}
+
+	traceID, err := trace.TraceIDFromHex(triggerRequest.TraceID)
+	if err != nil {
+		return fmt.Errorf("invalid traceID was received in TriggerRequest: %w", err)
+	}
+
+	response, err := triggerer.Trigger(ctx, triggerConfig, &agentTrigger.Options{
+		TraceID: traceID,
+		TestID:  id.ID(triggerRequest.TestID),
+	})
+
+	if err != nil {
+		return fmt.Errorf("could not trigger test: %w", err)
+	}
+
+	protoResponse := convertResponseToProtoResponse(triggerRequest, response)
+	err = w.client.SendTriggerResponse(ctx, protoResponse)
+	if err != nil {
+		return fmt.Errorf("could not send trigger response to server: %w", err)
+	}
+
+	return nil
+}
+
+func convertProtoToTrigger(pt *proto.Trigger) trigger.Trigger {
+	return trigger.Trigger{
+		Type:    trigger.TriggerType(pt.Type),
+		HTTP:    convertProtoHttpTriggerToHttpTrigger(pt.Http),
+		GRPC:    convertProtoGrpcTriggerToGrpcTrigger(pt.Grpc),
+		TraceID: convertProtoTraceIDTriggerToTraceIDTrigger(pt.TraceID),
+	}
+}
+
+func convertProtoHttpTriggerToHttpTrigger(pt *proto.HttpRequest) *trigger.HTTPRequest {
+	if pt == nil {
+		return nil
+	}
+
+	headers := make([]trigger.HTTPHeader, 0, len(pt.Headers))
+
+	for _, header := range pt.Headers {
+		headers = append(headers, trigger.HTTPHeader{Key: header.Key, Value: header.Value})
+	}
+
+	return &trigger.HTTPRequest{
```

Comment:
would it make sense to use `DeepCopy` instead of manually mapping fields?

Reply:
I thought about it. However, it would require adding the `dc` tag to some fields in the server's `trigger.Trigger` struct. I'm not convinced the server core should be aware of these details about the agent.